src/Main.elm

```diff
@@ -1,10 +1,571 @@
+            Dict.get parentId allNodesInDict
+
+        listOfParentChild : List { parent : Maybe Node, child : Node }
+        listOfParentChild =
+            allNodes
+                |> List.map (\node -> getParentChildNodes node)
+
+        -- This is flat tree with no parent-child relation yet
+        flatTree : Tree Node
+        flatTree =
+            allNodes
+                |> List.map Tree.singleton
+                |> Tree.tree Root
+    in
+    flatTreeToParentChildTree listOfParentChild (Just flatTree)
+
+
+
+-- This will order nodes in parent-child hierarchy
+
+
+flatTreeToParentChildTree : List { parent : Maybe Node, child : Node } -> Maybe (Tree Node) -> Tree Node
+flatTreeToParentChildTree listOfParentChild tree =
+    case listOfParentChild of
+        parentChild :: rest ->
+            case parentChild.parent of
+                Just parent ->
+                    flatTreeToParentChildTree rest (reorderNode { parent = parent, child = parentChild.child } tree)
+
+                Nothing ->
+                    flatTreeToParentChildTree rest tree
+
+        [] ->
+            Maybe.withDefault defaultTree tree
+
+
+reorderNode : { parent : Node, child : Node } -> Maybe (Tree Node) -> Maybe (Tree Node)
+reorderNode parentChild tree =
+    let
+        -- subTree is the child node
+        subTree : Maybe (Tree Node)
+        subTree =
+            findASubTree tree parentChild.child
+    in
+    -- tree is flat with no parent child hierarchy yet
+    tree
+        -- remove the child node from flat tree because we will insert it to parent soon
+        |> removeASubTree parentChild.child
+        -- and insert the subTree(child) to its parent tree
+        |> appendASubTreeAsChild subTree parentChild
+
+
+appendASubTreeAsChild : Maybe (Tree Node) -> { parent : Node, child : Node } -> Maybe (Tree Node) -> Maybe (Tree Node)
+appendASubTreeAsChild subTree parentChild tree =
+    let
+        appendTreeToTree tree_ =
+            Tree.appendChild (Maybe.withDefault defaultTree subTree) tree_
+    in
+    Zipper.fromTree (Maybe.withDefault defaultTree tree)
+        -- Find the parent node
+        |> Zipper.findFromRoot (\id -> id == parentChild.parent)
+        -- Insert subTree to the parent node
+        |> Maybe.map (Zipper.mapTree appendTreeToTree)
+        |> Maybe.map Zipper.toTree
+
+
+findASubTree : Maybe (Tree Node) -> Node -> Maybe (Tree Node)
+findASubTree tree subTreeNode =
+    Zipper.fromTree (Maybe.withDefault defaultTree tree)
+        |> Zipper.findFromRoot (\id -> id == subTreeNode)
+        |> Maybe.map Zipper.tree
+
+
+removeASubTree : Node -> Maybe (Tree Node) -> Maybe (Tree Node)
+removeASubTree node tree =
+    Zipper.fromTree (Maybe.withDefault defaultTree tree)
+        |> Zipper.findFromRoot (\id -> id == node)
+        |> Maybe.andThen Zipper.removeTree
+        |> Maybe.map Zipper.toTree
+
+
+defaultTree : Tree.Tree Node
+defaultTree =
+    Tree.singleton (File defaultAudience.id defaultAudience)
```

Comment:
The `defaultTree` seems crucial to have for your representation of the problem (`Maybe (Tree Node)`); I just want to suggest that it might be not needed at all.

I'd rather show an error to the user than show them a made-up data like 
```elm
 defaultAudienceFolder : AudienceFolder
 defaultAudienceFolder =
     { id = 0
     , name = ""
     , parent = Nothing
     }

 defaultAudience : Audience
 defaultAudience =
     { id = 0
     , name = ""
     , type_ = Curated
     , folder = Nothing
     }
```

Also there seems to be a lot of intermediate `Maybe.withDefault` unwrapping - it might be better to use `Maybe.map` and `Maybe.andThen` functions and stay "inside" the Maybe type for larger portion of the pipelines, instead of unwrapping to these made up values immediately after calling a function that creates a `Maybe`.